File: LeetCodes/minimum-number-of-arrows-to-burst-balloons/1-5-2023, 4_03_24 PM/minimum-number-of-arrows-to-burst-balloons.cpp
// Link: https://leetcode.com/problems/minimum-number-of-arrows-to-burst-balloons

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: minimum-number-of-arrows-to-burst-balloons
 *  RUNTIME: 1004 ms
 *  MEMORY: 89.8 MB
 *  DATE: 1-5-2023, 4:03:24 PM
 *
 */

// Solution:

class Solution {
    
public:
    int findMinArrowShots(vector<vector<int>>& points) {
        int n = points.size();
        if(n == 1) return 1;
        sort(points.begin(), points.end(), [](const vector<int> &x, const vector<int> &y) {return (x[1] < y[1]); });
        int i = 0, j = 0, count = 0;
        while(i < n) {
            if(count == 0 || points[i][0] > points[j][1]){
                count++;
                j = i;
            }
            i++;
        }
        return count;
    }
};

File: LeetCodes/minimum-number-of-arrows-to-burst-balloons/3-18-2024, 9_48_56 AM/minimum-number-of-arrows-to-burst-balloons.java
// Link: https://leetcode.com/problems/minimum-number-of-arrows-to-burst-balloons

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: minimum-number-of-arrows-to-burst-balloons
 *  RUNTIME: 54 ms
 *  MEMORY: 68.9 MB
 *  DATE: 3-18-2024, 9:48:56 AM
 *
 */

// Solution:

class Solution {
    public int findMinArrowShots(int[][] points) {
        Arrays.sort(points, (a, b)-> {
            if(a[0] == b[0]) 
                return Integer.compare(a[1], b[1]);
            return Integer.compare(a[0], b[0]);
        });

        int count = 1;
        long prevEnd = points[0][1];
        for(int[] arr: points) {
            if(arr[0] <= prevEnd) {
                if(arr[1] < prevEnd) {
                    prevEnd = arr[1];
                }
            } else {
                prevEnd = arr[1];
                count++;
            }
        }
        return count;
    }
}
